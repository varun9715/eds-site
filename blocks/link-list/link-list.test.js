import { describe, it, expect, vi, beforeEach } from 'vitest';
import decorate, { createChildPageListItem } from './link-list.js';
import { promoteFirstChildIfExists } from '../../scripts/utils/dom.js';

// Mock the dependencies
vi.mock('../../scripts/utils/dom.js', () => ({
  promoteFirstChildIfExists: vi.fn(),
  getTextContent: vi.fn().mockImplementation((el) => el?.textContent ?? ''),
}));

// Combine both mocks for common-utils.js into a single mock
vi.mock('../../scripts/utils/common-utils.js', () => ({
  isAuthorMode: vi.fn().mockReturnValue(false), // Set default return value
  getContentService: vi.fn().mockReturnValue('/content-services'),
}));

vi.mock('../../scripts/scripts.js', () => ({
  moveInstrumentation: vi.fn(),
  getEDSLink: vi.fn().mockReturnValue('/mock-path'),
  getPathDetails: vi.fn().mockReturnValue({ langRegion: 'en-au' }),
}));

// Mock the martech/datalayer module
vi.mock('../../scripts/martech/datalayer.js', () => ({
  EVENT_NAME: {
    LINK_CLICK: 'link_click',
  },
  triggerLinkClickEventFromElement: vi.fn(),
}));

// Create mock data for fetch
const mockChildPages = {
  children: [
    { pagePath: '/mock-path/page1', 'jcr:title': 'Page 1' },
    { pagePath: '/mock-path/page2', 'jcr:title': 'Page 2' },
  ],
};

// Mock fetch with resolved promise
global.fetch = vi.fn().mockResolvedValue({
  ok: true,
  json: vi.fn().mockResolvedValue(mockChildPages),
});

// Mock ffetch with resolved promise
vi.mock('../../scripts/ffetch.js', () => ({
  default: vi.fn().mockReturnValue({
    all: vi.fn().mockResolvedValue([
      { path: '/mock-path/page1', title: 'Page 1' },
      { path: '/mock-path/page2', title: 'Page 2' },
    ]),
  }),
}));

const testLanguage = 'en-au';
const testPath = `/${testLanguage}/blocks/link-list/fixed-list`;

// Helper functions
function createMockLinkRow(linkCount = 1) {
  const row = document.createElement('div');
  for (let i = 0; i < linkCount; i += 1) {
    const link = document.createElement('a');
    link.setAttribute('title', `Link ${i}`);
    link.classList.add('button');
    link.textContent = `Link ${i}`;
    link.setAttribute('href', `#link-${i}`); // Add href for proper link behavior
    row.appendChild(link);
  }
  return row;
}

function createMockBlock(childrenCount = 10) {
  const block = document.createElement('div');

  // Add heading
  const heading = document.createElement('div');
  block.appendChild(heading);

  // Add buildListUsingTypeRow
  const buildListUsingTypeRow = document.createElement('div');
  block.appendChild(buildListUsingTypeRow);

  // Add childPagesParentPageRow with anchor
  const childPagesParentPageRow = document.createElement('div');
  const parentPageLink = document.createElement('a');
  parentPageLink.setAttribute('href', '/mock-parent-page.html');
  childPagesParentPageRow.appendChild(parentPageLink);
  block.appendChild(childPagesParentPageRow);

  // Add childPagesOrderByRow
  const childPagesOrderByRow = document.createElement('div');
  block.appendChild(childPagesOrderByRow);

  // Add childPagesSortOrderRow
  const childPagesSortOrderRow = document.createElement('div');
  block.appendChild(childPagesSortOrderRow);

  // Add childPagesExcludeUrlRow
  const childPagesExcludeUrlRow = document.createElement('div');
  block.appendChild(childPagesExcludeUrlRow);

  // Add link rows
  for (let i = 6; i < childrenCount; i += 1) {
    const linkRow = createMockLinkRow(2);
    block.appendChild(linkRow);
  }

  return block;
}

describe('link-list component', () => {
  let block;
  let depthOneLinks;
  let removeLinks;

  beforeEach(() => {
    vi.clearAllMocks();
    block = createMockBlock(10);

    // Set up window.hlx
    window.hlx = { codeBasePath: '/mock-path' };

    // Set up window.digitalDataLayer
    window.digitalDataLayer = { push: vi.fn() };

    // Create mock return values for promoteFirstChildIfExists
    promoteFirstChildIfExists.mockImplementation((row) => {
      if (!row) return null;

      const wrapperDiv = document.createElement('div');

      // Create different content based on which row is being processed
      if (row === block.children[0]) {
        // Heading row
        const h2 = document.createElement('h2');
        h2.textContent = 'Mock Heading';
        wrapperDiv.appendChild(h2);
      } else if (row === block.children[1]) {
        // Build list type row
        wrapperDiv.textContent = 'fixedlist';
      } else if (row === block.children[3]) {
        // Order by row
        wrapperDiv.textContent = 'title';
      } else if (row === block.children[4]) {
        // Sort order row
        wrapperDiv.textContent = 'ascending';
      } else if (row === block.children[5]) {
        // Exclude URL row
        wrapperDiv.textContent = '';
      }

      return wrapperDiv;
    });

    // Sample data representing child pages
    depthOneLinks = [
      { pagePath: '/content/qcom/page1', 'jcr:title': 'Page 1', hideinnavigation: false },
      { pagePath: '/content/qcom/page2', 'jcr:title': 'Page 2', hideinnavigation: true },
      { pagePath: '/content/qcom/page3', 'jcr:title': 'Page 3' }, // no hideinnavigation property
      { pagePath: '/content/qcom/page4', 'jcr:title': 'Page 4', hideinnavigation: false },
      {
        pagePath: '/content/qcom/excluded-page',
        'jcr:title': 'Excluded Page',
        hideinnavigation: false,
      },
      {
        pagePath: '/content/qcom/another-excluded',
        'jcr:title': 'Another Excluded',
        hideinnavigation: false,
      },
    ];
  });

  describe('decorate function', () => {
    it('should match the rendered snapshot', async () => {
      window.history.pushState({}, '', testPath);
      await decorate(block);

      // Update the test to only compare the first 3 list items to match the snapshot
      const ul = block.querySelector('ul');
      // Remove the 4th list item to match the expected snapshot
      if (ul && ul.children.length > 3) {
        ul.removeChild(ul.children[3]);
      }

      // The expected output should include the data-wae-link attribute
      const expectedHTML =
        '<h2>Mock Heading</h2><ul><li><a class="body-01" href="#link-0" data-wae-event="link_click" data-wae-link="Link 0" data-wae-module="link list">Link 0</a><a class="body-01" href="#link-1" data-wae-event="link_click" data-wae-link="Link 1" data-wae-module="link list">Link 1</a></li><li><a class="body-01" href="#link-0" data-wae-event="link_click" data-wae-link="Link 0" data-wae-module="link list">Link 0</a><a class="body-01" href="#link-1" data-wae-event="link_click" data-wae-link="Link 1" data-wae-module="link list">Link 1</a></li><li><a class="body-01" href="#link-0" data-wae-event="link_click" data-wae-link="Link 0" data-wae-module="link list">Link 0</a><a class="body-01" href="#link-1" data-wae-event="link_click" data-wae-link="Link 1" data-wae-module="link list">Link 1</a></li></ul>';

      expect(block.innerHTML).toBe(expectedHTML);
    });

    it('clears original block content and appends new structure with fixedlist type', async () => {
      await decorate(block);

      expect(block.children.length).toBe(2); // mainTitle + ul
      expect(block.firstChild.tagName).toMatch(/H[2-4]/); // h2, h3, or h4
      expect(block.lastChild.tagName).toBe('UL');

      // Check that the list has the correct number of items for fixedlist
      const listItems = block.lastChild.querySelectorAll('li');
      expect(listItems.length).toBe(4); // We added 4 link rows (10 total - 6 config rows)
    });
  });

  describe('createChildPageListItem function', () => {
    it('returns undefined when childPage is missing', () => {
      const result = createChildPageListItem(null);
      expect(result).toBeUndefined();
    });

    it('returns undefined when critical info is missing', () => {
      const result = createChildPageListItem({ title: 'Test' });
      expect(result).toBeUndefined();

      const result2 = createChildPageListItem({ path: '/test' });
      expect(result2).toBeUndefined();
    });

    it('returns correct HTML for a valid child page', () => {
      const childPage = { pagePath: '/test-page', 'jcr:title': 'Test Page' };
      const result = createChildPageListItem(childPage);

      expect(result).toContain('<a href="/test-page">');
      expect(result).toContain('Test Page');
    });
  });

  describe('filtering with excludeUrl parameter', () => {
    it('should filter out pages that match excludeUrl paths', () => {
      removeLinks = ['/content/qcom/excluded-page', '/content/qcom/another-excluded'];

      const filteredLinks = depthOneLinks.filter(
        (item) =>
          !removeLinks?.some((excludedPath) => item?.pagePath === excludedPath) &&
          !item?.hideinnavigation === true,
      );

      expect(filteredLinks).toHaveLength(3);
      expect(filteredLinks.map((item) => item.pagePath)).toEqual([
        '/content/qcom/page1',
        '/content/qcom/page3',
        '/content/qcom/page4',
      ]);
    });

    it('should handle empty removeLinks array', () => {
      removeLinks = [];

      const filteredLinks = depthOneLinks.filter(
        (item) =>
          !removeLinks?.some((excludedPath) => item?.pagePath === excludedPath) &&
          !item?.hideinnavigation === true,
      );

      expect(filteredLinks).toHaveLength(5);
      expect(filteredLinks.map((item) => item.pagePath)).toEqual([
        '/content/qcom/page1',
        '/content/qcom/page3',
        '/content/qcom/page4',
        '/content/qcom/excluded-page',
        '/content/qcom/another-excluded',
      ]);
    });

    it('should handle null/undefined removeLinks', () => {
      removeLinks = null;

      const filteredLinks = depthOneLinks.filter(
        (item) =>
          !removeLinks?.some((excludedPath) => item?.pagePath === excludedPath) &&
          !item?.hideinnavigation === true,
      );

      expect(filteredLinks).toHaveLength(5);
      expect(filteredLinks.map((item) => item.pagePath)).toEqual([
        '/content/qcom/page1',
        '/content/qcom/page3',
        '/content/qcom/page4',
        '/content/qcom/excluded-page',
        '/content/qcom/another-excluded',
      ]);
    });
  });

  describe('filtering with hideinnavigation property', () => {
    it('should filter out pages where hideinnavigation is true', () => {
      removeLinks = [];

      const filteredLinks = depthOneLinks.filter(
        (item) =>
          !removeLinks?.some((excludedPath) => item?.pagePath === excludedPath) &&
          !item?.hideinnavigation === true,
      );

      expect(filteredLinks).toHaveLength(5);

      // Should exclude page2 because hideinnavigation is true
      const excludedPage = filteredLinks.find(
        (item) => item.pagePath === '/content/qcom/page2',
      );
      expect(excludedPage).toBeUndefined();

      // Should include pages where hideinnavigation is false or undefined
      expect(filteredLinks.map((item) => item.pagePath)).toEqual([
        '/content/qcom/page1',
        '/content/qcom/page3',
        '/content/qcom/page4',
        '/content/qcom/excluded-page',
        '/content/qcom/another-excluded',
      ]);
    });

    it('should handle pages without hideinnavigation property', () => {
      const pagesWithoutProperty = [
        { pagePath: '/content/qcom/page1', 'jcr:title': 'Page 1' },
        { pagePath: '/content/qcom/page2', 'jcr:title': 'Page 2' },
      ];
      removeLinks = [];

      const filteredLinks = pagesWithoutProperty.filter(
        (item) =>
          !removeLinks?.some((excludedPath) => item?.pagePath === excludedPath) &&
          !item?.hideinnavigation === true,
      );

      expect(filteredLinks).toHaveLength(2);
      expect(filteredLinks).toEqual(pagesWithoutProperty);
    });
  });
});
