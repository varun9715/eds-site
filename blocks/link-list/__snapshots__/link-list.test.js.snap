// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`link-list component > decorate function > should match the rendered snapshot 1`] = `"<h2>Mock Heading</h2><ul><li><a class="body-01" data-wae-event="link_click" data-wae-link="Link 0" data-wae-module="link list">Link 0</a><a class="body-01" data-wae-event="link_click" data-wae-link="Link 1" data-wae-module="link list">Link 1</a></li><li><a class="body-01" data-wae-event="link_click" data-wae-link="Link 0" data-wae-module="link list">Link 0</a><a class="body-01" data-wae-event="link_click" data-wae-link="Link 1" data-wae-module="link list">Link 1</a></li><li><a class="body-01" data-wae-event="link_click" data-wae-link="Link 0" data-wae-module="link list">Link 0</a><a class="body-01" data-wae-event="link_click" data-wae-link="Link 1" data-wae-module="link list">Link 1</a></li></ul>"`;
